elrond-codec-derive/src/lib.rs

```diff
@@ -0,0 +1,353 @@
+                        #(<#types>::dep_decode_or_exit(input, c.clone(), exit)),*
+                    )
+                }
+            }
+        };
+    }
+
+    gen.into()
+}
+
+// Top
+
+fn impl_top_encode_macro(ast: &syn::DeriveInput) -> TokenStream {
+    let name = &ast.ident;
+    let gen = match &ast.data {
+        syn::Data::Struct(_) => {
+            let (impl_generics, ty_generics, where_clause) = &ast.generics.split_for_impl();
+
+            quote! {
+                impl #impl_generics TopEncode for #name #ty_generics #where_clause {
+                    #[inline]
+                    fn top_encode<O: TopEncodeOutput>(&self, output: O) -> Result<(), EncodeError> {
+                        top_encode_from_nested(self, output)
+                    }
+                
+                    #[inline]
+                    fn top_encode_or_exit<O: TopEncodeOutput, ExitCtx: Clone>(
+                        &self,
+                        output: O,
+                        c: ExitCtx,
+                        exit: fn(ExitCtx, EncodeError) -> !,
+                    ) {
+                        top_encode_from_nested_or_exit(self, output, c, exit);
+                    }
+                }
+            }
+        },
+        syn::Data::Enum(_) => {
+            let idents = extract_field_names(&ast.data);
+            let value = 0..idents.len() as u8;
+            let name_repeated = std::iter::repeat(name);
+
+            quote! {
+                impl #name {
+                    pub fn to_u8(&self) -> u8 {
+                        match self {
+                            #(#name_repeated::#idents => #value,)*
+                        }
+                    }
+                }
+
+                impl TopEncode for #name {
+                    fn top_encode<O: TopEncodeOutput>(&self, output: O) -> Result<(), EncodeError> {
+                        self.to_u8().top_encode(output)
+                    }
+                
+                    fn top_encode_or_exit<O: TopEncodeOutput, ExitCtx: Clone>(
+                        &self,
+                        output: O,
+                        c: ExitCtx,
+                        exit: fn(ExitCtx, EncodeError) -> !,
+                    ) {
+                        self.to_u8().top_encode_or_exit(output, c, exit)
+                    }
+                }
+            }
+        },
+        syn::Data::Union(_) => panic!("Union not supported")
+    };
+
+    gen.into()
+}
+
+fn impl_top_decode_macro(ast: &syn::DeriveInput) -> TokenStream {
+    let name = &ast.ident;
+    let gen = match &ast.data {
+        syn::Data::Struct(_) => {
+            let (impl_generics, ty_generics, where_clause) = &ast.generics.split_for_impl();
+
+            quote! {
+                impl #impl_generics TopDecode for #name #ty_generics #where_clause {
+                    fn top_decode<I: TopDecodeInput>(input: I) -> Result<Self, DecodeError> {
+                        top_decode_from_nested(input)
```

Comment:
Long-term, we probably shouldn't assume NestedDecode is implemented, but for now it will be implemented for all structs.